core/src/main/java/org/apache/calcite/rel/externalize/RelJsonWriter.java

```diff
@@ -140,6 +142,17 @@ public RelWriter done(RelNode node) {
   public String asString() {
     final Map<String, Object> map = jsonBuilder.map();
     map.put("rels", relList);
-    return jsonBuilder.toJsonString(map);
+    AtomicInteger skipNormalize = null;
+    if (getDetailLevel() != SqlExplainLevel.DIGEST_ATTRIBUTES) {
+      skipNormalize = RexNode.DESCRIBE_WITHOUT_NORMALIZE.get();
+      skipNormalize.incrementAndGet();
+    }
+    try {
```

Comment:
This wrapper appears in many places, can we write a method to encapsulate that ?

Reply:
I would consider that to be on the hot path (especially in RexCall), so I would refrain from allocating extra objects there.

Adding a method like `withoutNormalize(Runnable)` won't make a huge readability difference, yet it would add overhead. The API is not intended for public use, so I'm inclined to leave it as is.

Reply:
I've extracted that to `class RexNode { ... public static Closeable skipNormalize()` which is to be used in try-with-resources.